server/src/main/java/com/tencent/rss/server/buffer/ShuffleBufferManager.java

```diff
@@ -293,14 +357,68 @@ public long getCapacity() {
   }
 
   @VisibleForTesting
-  void resetSize() {
+  public void resetSize() {
     usedMemory = new AtomicLong(0L);
     preAllocatedSize = new AtomicLong(0L);
     inFlushSize = new AtomicLong(0L);
   }
 
+  @VisibleForTesting
+  public Map<String, Map<Integer, AtomicLong>> getShuffleSizeMap() {
+    return shuffleSizeMap;
+  }
+
   public long getPreAllocatedSize() {
     return preAllocatedSize.get();
   }
 
+  // sort for shuffle according to data size, then pick properly data which will be flushed
+  private Map<String, Set<Integer>> pickFlushedShuffle() {
+    // create list for sort
+    List<Entry<String, AtomicLong>> sizeList = generateSizeList();
+    sizeList.sort((entry1, entry2) -> {
+      if (entry1.getValue().get() > entry2.getValue().get()) {
+        return -1;
+      }
+      return 1;
+    });
+    Map<String, Set<Integer>> pickedShuffle = Maps.newHashMap();
+    // The algorithm here is to flush data size > highWaterMark - lowWaterMark
+    // the remain data in buffer maybe more than lowWaterMark
+    // because shuffle server is still receiving data, but it should be ok
+    long expectedFlushSize = highWaterMark - lowWaterMark;
+    long pickedFlushSize = 0L;
+    for (Map.Entry<String, AtomicLong> entry : sizeList) {
+      long size = entry.getValue().get();
+      pickedFlushSize += size;
+      String appIdShuffleIdKey = entry.getKey();
+      addPickedShuffle(appIdShuffleIdKey, pickedShuffle);
+      if (pickedFlushSize > expectedFlushSize) {
+        LOG.info("Finish flush pick with {} bytes", pickedFlushSize);
+        break;
+      }
+    }
+    return pickedShuffle;
+  }
+
+  private List<Map.Entry<String, AtomicLong>> generateSizeList() {
+    Map<String, AtomicLong> sizeMap = Maps.newHashMap();
+    for (Map.Entry<String, Map<Integer, AtomicLong>> appEntry : shuffleSizeMap.entrySet()) {
+      String appId = appEntry.getKey();
+      for (Map.Entry<Integer, AtomicLong> shuffleEntry : appEntry.getValue().entrySet()) {
+        Integer shuffleId = shuffleEntry.getKey();
+        sizeMap.put(appId + KEY_SPLIT_CHAR + shuffleId, shuffleEntry.getValue());
```

Comment:
RssUtils.generateShuffleKey

Reply:
It's the only place using key as `appId + KEY_SPLIT_CHAR + shuffleId`, it can be wrapped as a method, but shouldn't be added to RssUtils until it is called from other place.

Reply:
We use appId + / + shuffleId as shufflekey in mulitstorage code.